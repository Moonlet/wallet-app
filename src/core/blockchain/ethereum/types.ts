export enum MethodSignature {
    INCREASE_ALLOWANCE = 'increaseAllowance(address,uint256)',
    DELEGATE = 'delegate(address,uint256)',
    UNDELEGATE = 'undelegate(address,uint256)',
    WITHDRAW_DELEGATED = 'withdrawDelegated(address,address)',
    TRANSFER = 'transfer(address,uint256)'
}
