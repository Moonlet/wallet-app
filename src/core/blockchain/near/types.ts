import BigNumber from 'bignumber.js';
import { PosBasicActionType } from '../types/token';

export enum NearTransactionActionType {
    TRANSFER = 'TRANSFER',
    FUNCTION_CALL = 'FUNCTION_CALL'
}

export enum NearFunctionCallMethods {
    DEPOSIT = 'deposit',
    STAKE = 'stake',
    UNSTAKE = 'unstake',
    WITHDRAW = 'withdraw'
}

export interface INearTransactionAction {
    type: NearTransactionActionType;
    params?: any[];
}

export interface INearTransactionAdditionalInfoType {
    currentBlockHash: string;
    actions: INearTransactionAction[];
    posAction?: PosBasicActionType;
    validatorName?: string;
}

export enum NearAccountType {
    DEFAULT = 'DEFAULT',
    CONTRACT = 'CONTRACT'
}

export interface INearAccount {
    address: string;
    name: string;
    amount: BigNumber;
    exists: boolean;
    valid: boolean;
    type?: NearAccountType;
}
