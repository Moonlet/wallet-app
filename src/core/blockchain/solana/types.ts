export enum SolanaTransactionInstructionType {
    TRANSFER = 'TRANSFER',
    CREATE_ACCOUNT_WITH_SEED = 'CREATE_ACCOUNT_WITH_SEED',
    DELEGATE_STAKE = 'DELEGATE_STAKE',
    SPLIT_STAKE = 'SPLIT_STAKE',
    UNSTAKE = 'UNSTAKE'
}

export interface ISolanaTransactionInstruction {
    type: SolanaTransactionInstructionType;
    instruction: any;
}

export interface IStakeAccountFormat {
    index: number;
    validatorId?: string;
    staked?: string;
    unstaked?: string;
    activating?: string;
    deactivating?: string;
}
