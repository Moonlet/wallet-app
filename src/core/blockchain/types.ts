export enum Blockchain {
    ETHEREUM = 'ETHEREUM',
    ZILLIQA = 'ZILLIQA'
}
