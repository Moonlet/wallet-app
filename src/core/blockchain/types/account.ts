import { IAccountState, ITokenState } from '../../../redux/wallets/state';
import { BigNumber } from 'bignumber.js';
import { ChainIdType, IFeeOptions } from '.';
import { PosBasicActionType } from './token';

export interface IBlockchainAccountUtils {
    getAccountDerivationPath(accountIndex: number): string;
    getPrivateKeyFromDerived(derivedKey: any): string;
    isValidChecksumAddress(address: string): boolean;
    isValidAddress(address: string): boolean;

    publicToAddress(publicKey: string): string;
    privateToAddress(privateKey: string): string;
    privateToPublic(privateKey: string): string;

    getAccountFromPrivateKey(privateKey: string, index: number): IAccountState;

    amountToStd(value: BigNumber | number | string, decimals: number): BigNumber;
    amountFromStd(value: BigNumber, decimals: number): BigNumber;
    convertUnit(value: BigNumber, from: string, to: string): BigNumber;

    availableAmount(
        account: IAccountState,
        value: string,
        token: ITokenState,
        chainId: ChainIdType,
        options: { feeOptions?: IFeeOptions; action?: PosBasicActionType | 'transfer' }
    ): Promise<string>;
}
