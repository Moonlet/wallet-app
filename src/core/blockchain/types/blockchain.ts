import {
    IBlockchainConfig,
    IBlockchainNetwork,
    IBlockchainAccountUtils,
    BlockchainGenericClient,
    IBlockchainTransactionUtils,
    ChainIdType
} from './';
import { GenericStats } from './stats';

export enum Blockchain {
    ETHEREUM = 'ETHEREUM',
    ZILLIQA = 'ZILLIQA',
    NEAR = 'NEAR',
    COSMOS = 'COSMOS',
    STELLAR = 'STELLAR'
}

export interface IBlockchain {
    config: IBlockchainConfig;
    networks: IBlockchainNetwork[];
    transaction: IBlockchainTransactionUtils;
    account: IBlockchainAccountUtils;
    Client: new (chainId) => BlockchainGenericClient;
    getStats: (chainId: ChainIdType) => GenericStats;
    getClient: (chainId: ChainIdType) => BlockchainGenericClient;
}
