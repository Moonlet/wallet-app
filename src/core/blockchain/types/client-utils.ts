import { IBlockchainTransaction } from './transaction';

export interface IClientUtils {
    getTransaction(hash: string, options?: { address?: string }): Promise<IBlockchainTransaction>;
}
