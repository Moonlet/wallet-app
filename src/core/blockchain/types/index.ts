export * from './blockchain';
export * from './client';
export * from './network';
export * from './transaction';
export * from './account';
export * from './config';
