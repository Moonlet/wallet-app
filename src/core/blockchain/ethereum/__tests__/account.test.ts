import { Ethereum } from '../';
import * as Util from 'ethereumjs-util';
import { Blockchain } from '../../types';

jest.mock('ethereumjs-util', () => {
    const util = jest.requireActual('ethereumjs-util');
    return {
        ...util,
        privateToPublic: jest.fn().mockReturnValue('PUBLIC_KEY'),
        privateToAddress: jest.fn().mockReturnValue('address'),
        toChecksumAddress: jest.fn().mockReturnValue('ADDRESS'),
        publicToAddress: jest.fn().mockReturnValue('address'),
        isValidChecksumAddress: jest.fn().mockReturnValue(true),
        isValidAddress: jest.fn().mockReturnValue(true)
    };
});

const clearMocks = () => {
    // @ts-ignore
    Util.privateToPublic.mockClear();
    // @ts-ignore
    Util.privateToAddress.mockClear();
    // @ts-ignore
    Util.toChecksumAddress.mockClear();
    // @ts-ignore
    Util.publicToAddress.mockClear();
    // @ts-ignore
    Util.isValidChecksumAddress.mockClear();
    // @ts-ignore
    Util.isValidAddress.mockClear();
};

describe('Ethereum account', () => {
    beforeEach(() => {
        clearMocks();
    });

    test('isValidChecksumAddress()', () => {
        expect(Ethereum.account.isValidChecksumAddress('ADDRESS')).toBe(true);
        expect(Util.isValidChecksumAddress).toBeCalledWith('ADDRESS');
    });

    test('isValidAddress()', () => {
        expect(Ethereum.account.isValidAddress('ADDRESS')).toBe(true);
        expect(Util.isValidAddress).toBeCalledWith('ADDRESS');
    });

    test('publicToAddress()', () => {
        expect(Ethereum.account.publicToAddress('PUBLIC_KEY')).toBe('ADDRESS');
        expect(Util.publicToAddress).toBeCalledWith(Buffer.from('PUBLIC_KEY', 'hex'));
        expect(Util.toChecksumAddress).toBeCalledWith('address');
    });

    test('privateToPublic()', () => {
        expect(Ethereum.account.privateToPublic('PRIVATE_KEY')).toBe('PUBLIC_KEY');
        expect(Util.privateToPublic).toBeCalledWith(Buffer.from('PRIVATE_KEY', 'hex'));
    });

    test('privateToAddress()', () => {
        expect(Ethereum.account.privateToAddress('PRIVATE_KEY')).toBe('ADDRESS');
        expect(Util.privateToAddress).toBeCalledWith(Buffer.from('PRIVATE_KEY', 'hex'));
        expect(Util.toChecksumAddress).toBeCalledWith('address');
    });

    test('getAccountFromPrivateKey()', () => {
        // valid private key
        const result = Ethereum.account.getAccountFromPrivateKey('PRIVATE_KEY', 1);
        expect(Util.privateToPublic).toBeCalledWith(Buffer.from('PRIVATE_KEY', 'hex'));
        expect(Util.privateToAddress).toBeCalledWith(Buffer.from('PRIVATE_KEY', 'hex'));
        expect(Util.toChecksumAddress).toBeCalledWith('address');

        expect(result).toEqual({
            index: 1,
            blockchain: Blockchain.ETHEREUM,
            address: 'ADDRESS',
            publicKey: 'PUBLIC_KEY'
        });

        // TODO: test with invalid privatekey, without mocks
    });
});
