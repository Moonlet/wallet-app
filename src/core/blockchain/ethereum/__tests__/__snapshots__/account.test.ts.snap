// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ethereum account getAccountFromPrivateKey() 1`] = `
Object {
  "address": "ADDRESS",
  "blockchain": "ETHEREUM",
  "index": 1,
  "publicKey": "PUBLIC_KEY",
  "selected": false,
  "tokens": Object {
    "ETH": Object {
      "active": true,
      "decimals": 18,
      "logo": Object {
        "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/eth.png",
      },
      "name": "Ethereum",
      "order": 0,
      "symbol": "ETH",
      "type": "NATIVE",
      "ui": Object {
        "decimals": 4,
        "tokenScreenComponent": "DEFAULT",
      },
      "units": Object {
        "ETH": "1000000000000000000",
        "GWEI": "1000000000",
        "WEI": "1",
      },
    },
  },
}
`;
